README.md
# TopNotchBottomBar

A beautiful and animated bottom navigation bar. also you are free to customize it way you want.

## Preview

![](https://github.com/sonal-sithara/top_notch_bottom_bar/blob/master/images/gif_1.gif)
![](https://github.com/sonal-sithara/top_notch_bottom_bar/blob/master/images/gif_2.gif)
![](https://github.com/sonal-sithara/top_notch_bottom_bar/blob/master/images/gif_3.gif)

## Customization

### TopNotchBottomBar

##### Required

- `items` - navigation items, required more than one item and less than six
- `onTap` - listen when a item is tapped it provide the selected item's index

##### Optional

- `backgroundColor` - the navigation bar's background color
- `height` - changes the bottom bar height
- `iconSize` - the item icon's size
- `showLabel` - the item's text
- `activeColor` - the item selected color
- `inActiveColor` - the item unselected color
- `showElevation` - if false the bottom bar elevation will be removed
- `animationDuration` - time duration for the notch and text animation
- `initialIndex` - set initial selected item
- `showCurvedBar` - if false the bottom bar curve will be removed
- `showCurvedBarAnimation` - if false the bottom bar curve animation will be removed

### TopNotchItem

##### Required

- `icon` - the icon of this item
- `name` - the text under the icon

##### Optional

- `color` - item color

## Getting Started

Add the dependency in `pubspec.yaml`:

```yaml
dependencies:
  top_notch_bottom_bar: ^0.1.0
```

## Basic Usage

Adding the widget

```dart
class _HomePageState extends State<HomePage> {
  static const TextStyle txtstyle = TextStyle(fontSize: 30);

  List<Widget> screens = [
    const Center(child: Text('Home', style: txtstyle)),
    const Center(child: Text('Favourite', style: txtstyle)),
    const Center(child: Text('Messages', style: txtstyle)),
    const Center(child: Text('Settings', style: txtstyle)),
  ];

  int index = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Top Notch Bottom Bar')),
      body: screens[index],
      bottomNavigationBar: TopNotchBottomBar(
          height: 56, // changes the bottom bar height -> default = 50
          onTap: ((value) => setState(() {
                index = value;
              })),
          items: [
            TopNotchItem(icon: const Icon(Icons.home), name: 'Home'),
            TopNotchItem(icon: const Icon(Icons.favorite), name: 'Favourite'),
            TopNotchItem(icon: const Icon(Icons.message), name: 'Messages'),
            TopNotchItem(icon: const Icon(Icons.settings), name: 'Settings'),
          ]),
    );
  }
}
```


CHANGELOG.md
## 0.0.1

- Initial beta release.

## 0.1.0

- Stable release.


lib/src/top_notch_bottom_bar.dart
import 'package:flutter/material.dart';
import 'constants/text_style.dart';
import 'data/top_notch_bottom_bar_item.dart';
import 'widgets/custom_paint.dart';

class TopNotchBottomBar extends StatefulWidget {
  final List<TopNotchItem> items;
  final bool? showLabel;
  final double? height;
  final double? iconSize;
  final Color? activeColor;
  final Color? inActiveColor;
  final int? initialIndex;
  final ValueChanged<int> onTap;
  final bool? showElevation;
  final Color? backgroundColor;
  final Duration? animationDuration;
  final bool? showCurvedBar;
  final bool? showCurvedBarAnimation;

  const TopNotchBottomBar({
    Key? key,
    required this.items,
    required this.onTap,
    this.initialIndex = 0,
    this.height = 50,
    this.animationDuration = const Duration(milliseconds: 400),
    this.iconSize,
    this.backgroundColor = Colors.white,
    this.showElevation = true,
    this.activeColor = const Color.fromARGB(255, 39, 153, 247),
    this.inActiveColor = const Color.fromARGB(255, 150, 150, 150),
    this.showLabel = true,
    this.showCurvedBar = true,
    this.showCurvedBarAnimation = true,
  }) : super(key: key);

  @override
  State<TopNotchBottomBar> createState() => _TopNotchBottomBarState();
}

class _TopNotchBottomBarState extends State<TopNotchBottomBar> {
  int? index = 0;

  @override
  void initState() {
    super.initState();
    index = widget.initialIndex;
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      height: widget.height,
      duration: widget.animationDuration!,
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        boxShadow: [
          if (widget.showElevation!)
            BoxShadow(
              color: Colors.grey.withOpacity(0.5),
              spreadRadius: 2,
              blurRadius: 7,
              offset: const Offset(0, 3), // changes position of shadow
            ),
        ],
        borderRadius: widget.showCurvedBar!
            ? BorderRadius.only(
                topLeft: widget.showCurvedBarAnimation!
                    ? Radius.circular(index == 0 ? 0.0 : 20.0)
                    : const Radius.circular(20.0),
                topRight: widget.showCurvedBarAnimation!
                    ? Radius.circular(
                        index == widget.items.length - 1 ? 0.0 : 20.0)
                    : const Radius.circular(20.0),
              )
            : const BorderRadius.all(Radius.zero),
      ),
      child: Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        for (int i = 0; i < widget.items.length; i++)
          Expanded(
            child: GestureDetector(
              onTap: () {
                setState(() => index = i);
                widget.onTap(index!);
              },
              child: iconBtn(i),
            ),
          ),
      ]),
    );
  }

  Container iconBtn(int i) {
    bool isActive = index == i;
    var height = isActive ? 10.0 : 0.0;
    var width = isActive ? 60.0 : 0.0;
    return Container(
      color: Colors.transparent,
      child: Stack(
        children: [
          Align(
            alignment: Alignment.topCenter,
            child: AnimatedContainer(
              height: height,
              width: width,
              duration: widget.animationDuration!,
              curve: Curves.easeIn,
              child: isActive
                  ? CustomPaint(
                      painter: ButtonNotch(activeColor: widget.activeColor!),
                    )
                  : const SizedBox(),
            ),
          ),
          Positioned.fill(
              child: Align(
            alignment: Alignment.center,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Icon(
                  widget.items[i].icon.icon,
                  color: isActive
                      ? widget.activeColor
                      : widget.items[i].color ?? widget.inActiveColor,
                  size: widget.iconSize,
                ),
                if (widget.showLabel!)
                  Padding(
                    padding: const EdgeInsets.only(top: 2),
                    child: AnimatedDefaultTextStyle(
                      duration: widget.animationDuration!,
                      style: isActive
                          ? textStyle.copyWith(color: widget.activeColor)
                          : textStyle.copyWith(
                              color: widget.items[i].color ??
                                  widget.inActiveColor),
                      child: Text(
                        widget.items[i].name,
                      ),
                    ),
                  ),
              ],
            ),
          ))
        ],
      ),
    );
  }
}


lib/src/widgets/custom_paint.dart
import 'package:flutter/cupertino.dart';

class ButtonNotch extends CustomPainter {
  Color activeColor;

  ButtonNotch({required this.activeColor});

  @override
  void paint(Canvas canvas, Size size) {
    // var dotPoint = Offset(size.width / 2, -4);

    var paint_1 = Paint()
      ..color = activeColor
      ..style = PaintingStyle.fill;

    // var paint_2 = Paint()
    //   ..color = Colors.white
    //   ..style = PaintingStyle.fill;

    var path = Path();

    path.moveTo(0, 0);
    path.quadraticBezierTo(7.5, 0, 12, 5);
    path.quadraticBezierTo(size.width / 2, size.height / 2, size.width - 12, 5);
    path.quadraticBezierTo(size.width - 7.5, 0, size.width, 0);
    path.close();
    canvas.drawPath(path, paint_1);
    // canvas.drawCircle(dotPoint, 4, paint_2);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}


lib/src/animated_bottom_nav_bar.dart
import 'package:flutter/material.dart';
import 'constants/text_style.dart';
import 'data/animated_bottom_nav_bar_item.dart';
import 'widgets/custom_paint.dart';

class TopNotchBottomBar extends StatefulWidget {
  final List<TopNotchItem> items;
  final bool? showLabel;
  final double? height;
  final double? iconSize;
  final Color? activeColor;
  final Color? inActiveColor;
  final int? initialIndex;
  final ValueChanged<int> onTap;
  final bool? showElevation;
  final Color? backgroundColor;
  final Duration? animationDuration;

  const TopNotchBottomBar({
    Key? key,
    required this.items,
    required this.onTap,
    this.initialIndex = 0,
    this.height = 50,
    this.animationDuration = const Duration(milliseconds: 400),
    this.iconSize,
    this.backgroundColor = Colors.white,
    this.showElevation = true,
    this.activeColor = const Color.fromARGB(255, 39, 153, 247),
    this.inActiveColor = const Color.fromARGB(255, 150, 150, 150),
    this.showLabel = true,
  }) : super(key: key);

  @override
  State<TopNotchBottomBar> createState() => _TopNotchBottomBarState();
}

class _TopNotchBottomBarState extends State<TopNotchBottomBar> {
  int? index = 0;

  @override
  void initState() {
    super.initState();
    index = widget.initialIndex;
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      height: widget.height,
      duration: widget.animationDuration!,
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        boxShadow: [
          if (widget.showElevation!)
            BoxShadow(
              color: Colors.grey.withOpacity(0.5),
              spreadRadius: 2,
              blurRadius: 7,
              offset: const Offset(0, 3), // changes position of shadow
            ),
        ],
        borderRadius: BorderRadius.only(
          topLeft: Radius.circular(index == 0 ? 0.0 : 20.0),
          topRight:
              Radius.circular(index == widget.items.length - 1 ? 0.0 : 20.0),
        ),
      ),
      child: Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
        for (int i = 0; i < widget.items.length; i++)
          Expanded(
            child: GestureDetector(
              onTap: () {
                setState(() => index = i);
                widget.onTap(index!);
              },
              child: iconBtn(i),
            ),
          ),
      ]),
    );
  }

  Container iconBtn(int i) {
    bool isActive = index == i;
    var height = isActive ? 10.0 : 0.0;
    var width = isActive ? 60.0 : 0.0;
    return Container(
      color: Colors.transparent,
      child: Stack(
        children: [
          Align(
            alignment: Alignment.topCenter,
            child: AnimatedContainer(
              height: height,
              width: width,
              duration: widget.animationDuration!,
              curve: Curves.easeIn,
              child: isActive
                  ? CustomPaint(
                      painter: ButtonNotch(activeColor: widget.activeColor!),
                    )
                  : const SizedBox(),
            ),
          ),
          Positioned.fill(
              child: Align(
            alignment: Alignment.center,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Icon(
                  widget.items[i].icon.icon,
                  color: isActive
                      ? widget.activeColor
                      : widget.items[i].color ?? widget.inActiveColor,
                  size: widget.iconSize,
                ),
                if (widget.showLabel!)
                  Padding(
                    padding: const EdgeInsets.only(top: 2),
                    child: AnimatedDefaultTextStyle(
                      duration: widget.animationDuration!,
                      style: isActive
                          ? textStyle.copyWith(color: widget.activeColor)
                          : textStyle.copyWith(
                              color: widget.items[i].color ??
                                  widget.inActiveColor),
                      child: Text(
                        widget.items[i].name,
                      ),
                    ),
                  ),
              ],
            ),
          ))
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:top_notch_bottom_bar/top_notch_bottom_bar.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      debugShowCheckedModeBanner: false,
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  static const TextStyle txtstyle = TextStyle(fontSize: 30);

  List<Widget> screens = [
    const Center(child: Text('Home', style: txtstyle)),
    const Center(child: Text('Favourite', style: txtstyle)),
    const Center(child: Text('Messages', style: txtstyle)),
    const Center(child: Text('Settings', style: txtstyle)),
  ];

  int index = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Top Notch Bottom Bar'),
        backgroundColor: Colors.cyan,
      ),
      body: screens[index],
      bottomNavigationBar: TopNotchBottomBar(
          backgroundColor: Colors.cyan,
          inActiveColor: Colors.black38,
          activeColor: Colors.white,
          height: 56, // changes the bottom bar height -> default = 50
          onTap: ((value) => setState(() {
                index = value;
              })),
          items: [
            TopNotchItem(icon: const Icon(Icons.home), name: 'Home'),
            TopNotchItem(icon: const Icon(Icons.favorite), name: 'Favourite'),
            TopNotchItem(icon: const Icon(Icons.message), name: 'Messages'),
            TopNotchItem(icon: const Icon(Icons.settings), name: 'Settings'),
          ]),
    );
  }
}
